sklearn_numba_dpex/kmeans/kernels.py

```diff
@@ -0,0 +1,727 @@
+
+    inf = dtype(math.inf)
+    zero = dtype(0.0)
+    one = dtype(1.0)
+
+    n_cluster_items = n_clusters * (n_features + 1)
+    n_cluster_bytes = 4 * n_cluster_items
+    # TODO: control that this value is not higher than the number of sub-groups of size
+    # preferred_work_group_size_multiple that can effectively run concurrently. We
+    # should fetch this information and apply it here.
+    n_centroids_private_copies = int(
+        (global_mem_cache_size * centroids_private_copies_max_cache_occupancy)
+        // n_cluster_bytes
+    )
+
+    # fmt: off
+    @dpex.kernel
+    def fused_kmeans(
+        X_t,                                # IN    (n_features, n_samples)
+        current_centroids_t,                # IN    (n_features, n_clusters)
+        centroids_half_l2_norm,             # IN    (n_clusters,)
+        per_sample_pseudo_inertia,          # OUT   (n_samples,)
+        new_centroids_t_private_copies,     # OUT   (n_private_copies, n_features, n_clusters)
+        cluster_sizes_private_copies,       # OUT   (n_private_copies, n_clusters)
+    ):
+    # fmt: on
+        """One full iteration of LLoyd's k-means.
+
+        The kernel is meant to be invoked on a 1D grid spanning the data samples of
+        the training set in parallel.
+
+        Each work-item will assign that sample to its nearest centroid and accumulate
+        the feature values of the data samples into the new version of the centroid
+        for the next iteration. The centroid assignment and centroid update steps are
+        performed in a single fused kernel to avoid introducing a large intermediate
+        label assignment array to be re-read from global memory before performing the
+        centroid update step.
+
+        To avoid atomic contention when performing the centroid updates concurrently
+        for different data samples, we use private copies of the new centroid array in
+        global memory. Those copies are meant to be re-reduced afterwards.
+
+        The distance is the euclidean distance. Note that it is not necessary to
+        compute the exact value to find the closest centroid. Indeed, minimizing
+            |x-y|^2 = |x|^2 - 2<x.y> + |y|^2
+        over y amounts to minimizing
+            (1/2)y^2 - xy .
+        Moreover the value (1/2)y^2 has been pre-computed in the array
+        centroids_half_l2_norm to reduce the overall number of floating point
+        operations in the kernel.
+        """
+        sample_idx = dpex.get_global_id(0)
+        local_work_id = dpex.get_local_id(0)
+
+        # This array in shared memory is used as a sliding array over values of
+        # current_centroids_t. During each iteration in the inner loop, a new one is 
+        # loaded and used by all work items in the work group to compute partial 
+        # results. The array slides over the features in the outer loop, and over the 
+        # samples in the inner loop.
```

Comment:
Thanks your for commenting this tensor's purpose and usage and all the other ones' bellow. :100: